test.sh

```diff
@@ -1,51 +1,92 @@
 #!/bin/bash -e
-# test suite...
-echo running test.sh
-echo "ENV:"
-env
+# runs all (or selected) test suite(s) in test/ and aggregates results
+# Usage:
+#	./test.sh
+#	./test.sh gofmt
 
-failures=''
-function run-test()
+# library of utility functions
+# shellcheck disable=SC1091
+. test/util.sh
+
+# allow specifying a single testsuite to run
+testsuite="$1"
+
+# print environment when running all testsuites
+test -z "$testsuite" && (echo "ENV:"; env; echo; )
+
+# run a test and record failures
+function run-testsuite()
 {
-	$@ || failures=$( [ -n "$failures" ] && echo "$failures\\n$@" || echo "$@" )
+	testname="$(basename "$1" .sh)"
+	# if not running all test or this test is not explicitly selected, skip it
+	if test -z "$testsuite" || test "test-$testsuite" = "$testname";then
+		$@ || failures=$( [ -n "$failures" ] && echo "$failures\\n$@" || echo "$@" )
+	fi
 }
 
-# ensure there is no trailing whitespace or other whitespace errors
-run-test git diff-tree --check $(git hash-object -t tree /dev/null) HEAD
+# only run test if it is explicitly selected, otherwise report it is skipped
+function skip-testsuite()
+{
+	testname=$(basename "$1" .sh)
+	# show skip message only when running full suite
+	if test -z "$testsuite";then
+		echo skipping "$@" "($REASON)"
+		blue SKIP
```

Comment:
colours don't exist in this commit yet.

Reply:
I've missed some when splitting up the commits, i'll correct this and the others